lib/AST/ASTContext.cpp

```diff
@@ -434,10 +434,31 @@ static ModuleDecl *createBuiltinModule(ASTContext &ctx) {
   return M;
 }
 
+inline ASTContext::Implementation &ASTContext::getImpl() const {
+  auto end = reinterpret_cast<void*>(const_cast<ASTContext*>(this) + 1);
+  auto impl = llvm::alignAddr(end, alignof(Implementation));
+  return *reinterpret_cast<Implementation*>(impl);
+}
+
+ASTContext *ASTContext::get(LangOptions &langOpts,
+                            SearchPathOptions &SearchPathOpts,
+                            SourceManager &SourceMgr,
+                            DiagnosticEngine &Diags) {
+  // If more than two data structures are concatentated, then the aggregate
+  // size math needs to become more complicated due to per-struct alignment
+  // constraints.
+  auto align = std::max(alignof(ASTContext), alignof(Implementation));
+  auto size = llvm::alignTo(sizeof(ASTContext) + sizeof(Implementation), align);
+  auto mem = AlignedAlloc(size, align);
```

Comment:
Unfortunately I think this is causing ASAN problems. The returned pointer is put into a `unique_ptr` at

https://github.com/apple/swift/blob/9cd07e9db30b7fe30ea5fcec422a40ea9bfd4995/lib/Frontend/Frontend.cpp#L152-L154

and then destroyed with `unique_ptr::reset` (which seems to use `delete` internally):

https://github.com/apple/swift/blob/9cd07e9db30b7fe30ea5fcec422a40ea9bfd4995/lib/Frontend/Frontend.cpp#L887

[ASAN doesn't like the mismatched allocators](https://ci.swift.org/job/oss-swift-incremental-ASAN-RA-ubuntu-16_04/3229):

```
15:25:33 ==14342==ERROR: AddressSanitizer: alloc-dealloc-mismatch (malloc vs operator delete) on 0x621000016900
15:25:33     #0 0x6f3e60 in operator delete(void*) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x6f3e60)
15:25:33     #1 0x2b6cd55 in swift::CompilerInstance::freeASTContext() (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x2b6cd55)
15:25:33     #2 0x7ebbe0 in performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x7ebbe0)
15:25:33     #3 0x7de4bf in swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x7de4bf)
15:25:33     #4 0x6fd1e1 in main (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x6fd1e1)
15:25:33     #5 0x7f3969e6882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
15:25:33     #6 0x622148 in _start (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x622148)
15:25:33 
15:25:33 0x621000016900 is located 0 bytes inside of 4680-byte region [0x621000016900,0x621000017b48)
15:25:33 allocated by thread T0 here:
15:25:33     #0 0x6c2b60 in posix_memalign (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x6c2b60)
15:25:33     #1 0x3cba8e7 in swift::ASTContext::get(swift::LangOptions&, swift::SearchPathOptions&, swift::SourceManager&, swift::DiagnosticEngine&) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x3cba8e7)
15:25:33     #2 0x2b61306 in swift::CompilerInstance::setup(swift::CompilerInvocation const&) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x2b61306)
15:25:33     #3 0x7ddc83 in swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x7ddc83)
15:25:33     #4 0x6fd1e1 in main (/home/buildnode/jenkins/workspace/oss-swift-incremental-ASAN-RA-ubuntu-16_04/buildbot_incremental_asan/swift-linux-x86_64/bin/swift+0x6fd1e1)
15:25:33     #5 0x7f3969e6882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
```

Reply:
#16537